### PR TITLE
Add k8s API error handling utility types and functions

### PR DESCRIPTION
**Pull request by axbarsan**
This PR adds a few utility types and functions for dealing with k8s API errors. It also changes the structure of the k8s api service types and functions, to one that I believe that makes more sense.

It also illustrates how all these changes work by providing an example:
- Fetching a `capiv1alpha3.Cluster`'s infrastructure ref (provider specific resource), and displaying some info for it. Currently, just `capzv1alpha3.AzureCluster` is handled.

### CODE REVIEW

File: src/components/MAPI/clusters/ClusterListItemDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+const ClusterListItemDetails: React.FC<IClusterListItemDetailsProps> = ({
+  cluster,
+}) => {
+  const client = useHttpClient();
+  const user = useSelector(getLoggedInUser);
+
+  // Use type intersection for multiple cluster types.
+  const { data, error } = useSWR<capzv1alpha3.IAzureCluster, GenericResponse>(
```
Comment (axbarsan) [inline src/components/MAPI/clusters/ClusterListItemDetails.tsx:24]: For multiple providers we can have

```suggestion
  const { data, error } = useSWR<capzv1alpha3.IAzureCluster | capav1alpha3.IAWSCluster | ..., GenericResponse>(
```

File: src/components/MAPI/clusters/ClusterListItemDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    capiv1alpha3.getClusterInfraRefKey(user, cluster),
+    capiv1alpha3.getClusterInfraRef(client, user!, cluster)
+  );
+
+  if (
+    metav1.isStatusError(error?.data, metav1.K8sStatusErrorReasons.NotFound)
```
Comment (axbarsan) [inline src/components/MAPI/clusters/ClusterListItemDetails.tsx:30]: How to check if an error is of a specific K8s API error type. In this block the error type will now get the reason and error status code.

image

File: src/components/MAPI/clusters/ClusterListItemDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    );
+  }
+
+  if (data?.kind === capzv1alpha3.AzureCluster) {
+    return (
+      <Box key='details'>
+        <Text>Region: {data.spec.location}</Text>
+      </Box>
+    );
+  }
```
Comment (axbarsan) [inline src/components/MAPI/clusters/ClusterListItemDetails.tsx:53]: Can have multiple blocks for multiple types of resources

File: src/model/services/mapi/capiv1alpha3/getClusterInfraRef.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    switch (infrastructureRef.kind) {
+      case capzv1alpha3.AzureCluster:
+        return capzv1alpha3.getAzureClusterByName(
+          client,
+          user,
+          cluster.metadata.namespace,
+          infrastructureRef.name
+        )();
+
```
Comment (axbarsan) [inline src/model/services/mapi/capiv1alpha3/getClusterInfraRef.ts:28]: Just add another case for handling a different type, same as in the key function below

File: src/model/services/mapi/capiv1alpha3/getClusterInfraRef.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    switch (infrastructureRef.kind) {
+      case capzv1alpha3.AzureCluster:
+        return capzv1alpha3.getAzureClusterByName(
+          client,
+          user,
+          cluster.metadata.namespace,
+          infrastructureRef.name
+        )();
```
Comment (axbarsan) [inline src/model/services/mapi/capiv1alpha3/getClusterInfraRef.ts:27]: Just add another case for handling a different type, same as in the key function below

**Review by oponder — APPROVED** [review]
LGTM!

**Pull request merged**